README.md
# KC11B04 Driver &emsp; [![Latest Version]][crates.io] [![Docs Badge]][docs]

[Latest Version]: https://img.shields.io/crates/v/kc11b04.svg
[crates.io]: https://crates.io/crates/kc11b04
[Docs Badge]: https://docs.rs/kc11b04/badge.svg
[docs]: https://docs.rs/kc11b04

Driver in Rust for the KC11B04 4-button analog keypad.

A 3-pin module using a voltage divider circuit and analog pin to sense a key press.

![KC11B04 image][__link0]


## Wiring notes

The voltage on the `VCC` pin should correspond with the reference voltage (maximum value) your ADC is configured to read.

For example the default on an Arduino Uno is that it matches the operating voltage of 5V. But the ADC can be configured to use an external reference with the `AREF` pin. In this case `VCC` should match `AREF`.


## Example


```rust
use kc11b04::{Key, KC11B04, MAP_10BIT};

let mut adc = /* Configure your ADC using its HAL */
let mut ad_pin = /* Set the pin connected to AD as analog input */

// Create the keypad driver. Taking ownership of the pin,
// providing a map that matches the resolution of your ADC.
let mut keypad = KC11B04::new(ad_pin, MAP_10BIT);

// Somewhere within loop { }
// Read current key state.
match keypad
	.key_state(&mut adc)
	.expect("Problem reading ADC channel")
{
	Some(Key::K4) => { /* K4 key being pressed */ }
	Some(Key::K3) => { /* K3 key being pressed */ }
	Some(Key::K2) => { /* K2 key being pressed */ }
	Some(Key::K1) => { /* K1 key being pressed */ }
	None => { /* Either nothing, or multiple keys pressed */ }
}
```


## MSRV & Stability

This crate compiles on stable Rust `1.60` and up. It *might* still work on older rust versions, but this isn’t ensured. Upgrading the MSRV is considered SemVer breaking.


 [__link0]: docs/KC11B04.webp


CHANGELOG.md
# Changelog

## \[0.3.0]

- [`2fc97fa`](https://github.com/Beanow/kc11b04-rs/commit/2fc97fadb95830a9b61aa38cd212dfc19dc09300)([#10](https://github.com/Beanow/kc11b04-rs/pull/10)) **Breaking**: support being generic over `Word` instead of assuming `u16` values from the ADC.

  This changes the type from `KeyMap` to `KeyMap<Word>` and adds a type as the first argument to the `map_from_max!` macro. For example:

  ```rust
  /// Signed 11bit map. Meaning 10 bits of resolution in the positive voltage range.
  const MAP_11BIT_SIGNED: KeyMap<i16> = map_from_max!(i16, 1023);
  ```

## \[0.2.1]

- [`4e15e28`](https://github.com/Beanow/kc11b04-rs/commit/4e15e28d236994c1513c84216757c8fb67829806)([#8](https://github.com/Beanow/kc11b04-rs/pull/8)) Include doc images in package which should fix docs.rs builds.

## \[0.2.0]

- [`1c3c5f5`](https://github.com/Beanow/kc11b04-rs/commit/1c3c5f512f2a8e6f637012ffdbd6742b213cc888)([#7](https://github.com/Beanow/kc11b04-rs/pull/7)) Custom maps can now be created.

  Either using the macro `map_from_max!` or constructing a `KeyMap` manually.
  See [`kc11b04::mapping`](https://docs.rs/kc11b04/0.2.1/kc11b04/mapping/index.html) docs.

- [`4e43c76`](https://github.com/Beanow/kc11b04-rs/commit/4e43c7604ab655606ff3343a40e00afcb9922469)([#5](https://github.com/Beanow/kc11b04-rs/pull/5)) Expand documentation with examples and images.

## \[0.1.0]

- [`f6a2ab0`](https://github.com/Beanow/kc11b04-rs/commit/f6a2ab0e341048e9d4ef19034f4af3c2c8092a53) Initial release of the crate :tada:


src/map.rs
// Copyright 2023-2023 Robin van Boven
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use crate::Key;

/// Maps keys to their expected ADC readings.
///
/// Note: this depends on the ADC resolution, so you may want to use a constant for this such as [`MAP_10BIT`].
#[cfg_attr(feature = "defmt-0-3", derive(defmt::Format))]
#[derive(Debug)]
pub struct KeyMap {
	k1: u16,
	k2: u16,
	k3: u16,
	k4: u16,
	margin: u16,
}

impl KeyMap {
	/// Takes an ADC reading and finds whether it's in the expected range of a key.
	///
	/// Will be [`None`] when no key is pressed, but also for some simultaneous key combinations.
	pub const fn key_from_reading(&self, val: u16) -> Option<Key> {
		match val {
			v if v > self.k1 - self.margin && v < self.k1 + self.margin => Some(Key::K1),
			v if v > self.k2 - self.margin && v < self.k2 + self.margin => Some(Key::K2),
			v if v > self.k3 - self.margin && v < self.k3 + self.margin => Some(Key::K3),
			v if v > self.k4 - self.margin => Some(Key::K4),
			_ => None,
		}
	}
}

/*
	Voltage divider with equivalent parallel resistors.
	NONE = AD is floating
	K1 = 3000 : ~1960 ... 1/((1/2000) + (1/100000))
	K2 = 2000 : ~2912 ... 1/((1/3000) + (1/100000))
	K3 = 1000 : ~3846 ... 1/((1/4000) + (1/100000))
	K4 = 0 : 100K
*/
const K1_F: f32 = 0.3952569;
const K2_F: f32 = 0.5928853;
const K3_F: f32 = 0.7936507;

/// [`KeyMap`] for 10bit ADCs with a maximum reading of `1023`.
pub const MAP_10BIT: KeyMap = {
	let max = 1023;
	let margin = 0.03;
	KeyMap {
		k1: (max as f32 * K1_F) as u16,
		k2: (max as f32 * K2_F) as u16,
		k3: (max as f32 * K3_F) as u16,
		k4: max,
		margin: (max as f32 * margin) as u16,
	}
};

#[test]
fn read_10bit_samples() {
	let map = MAP_10BIT;
	assert_eq!(
		(
			map.key_from_reading(0),
			map.key_from_reading(300),
			map.key_from_reading(400),
			map.key_from_reading(500),
			map.key_from_reading(600),
			map.key_from_reading(700),
			map.key_from_reading(800),
			map.key_from_reading(900),
			map.key_from_reading(1023)
		),
		(
			None,
			None,
			Some(Key::K1),
			None,
			Some(Key::K2),
			None,
			Some(Key::K3),
			None,
			Some(Key::K4)
		)
	);
}


src/driver.rs
// Copyright 2023-2023 Robin van Boven
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use core::{
	marker::PhantomData,
	ops::{Add, Sub},
};
use embedded_hal::adc::{Channel, OneShot};

use crate::{Key, KeyMap};

/// KC11B04 analog keypad driver. Constructed with [`KC11B04::new`].
pub struct KC11B04<Pin, ADC, Word> {
	pin: Pin,
	map: KeyMap<Word>,
	_adc: PhantomData<ADC>,
}

type Error<Adc, ADC, Word, Pin> = nb::Error<<Adc as OneShot<ADC, Word, Pin>>::Error>;

impl<Pin, ADC, Word> KC11B04<Pin, ADC, Word>
where
	Pin: Channel<ADC>,
	Word: Copy + Add<Output = Word> + Sub<Output = Word> + Ord,
{
	/// Create a [`KC11B04`] instance for the given ADC pin / channel and mapping.
	///
	/// ```rust
	/// # use embedded_hal_mock::adc::*;
	/// # use kc11b04::{Key, KC11B04};
	/// #
	/// # let mut adc = Mock::new(&[Transaction::read(0, 1023)]);
	/// # let analog_pin = MockChan0;
	/// #
	/// // Create the keypad driver. Taking ownership of the pin,
	/// // providing a map that matches the resolution of your ADC.
	/// let mut keypad = KC11B04::new(analog_pin, kc11b04::MAP_10BIT);
	/// ```
	pub fn new(pin: Pin, map: KeyMap<Word>) -> Self
	where
		Pin: Channel<ADC>,
	{
		Self {
			pin,
			map,
			_adc: PhantomData,
		}
	}

	/// Takes an ADC reading and finds whether a key is currently being pressed.
	///
	/// Will be [`None`] when no key is pressed, but also for some simultaneous key combinations.
	pub fn key_state<Adc>(
		&mut self,
		adc: &mut Adc,
	) -> Result<Option<Key>, Error<Adc, ADC, Word, Pin>>
	where
		Adc: OneShot<ADC, Word, Pin>,
	{
		let val = adc.read(&mut self.pin)?;
		Ok(self.map.key_from_reading(val))
	}
}

#[cfg(test)]
mod test {
	use crate::{Key, KC11B04, MAP_10BIT};
	use embedded_hal::adc::Channel;
	use embedded_hal_mock::adc::{Mock, MockChan0, Transaction};

	#[test]
	fn reads_given_channel() {
		use MockChan0 as PIN;

		let expected = [
			Transaction::read(PIN::channel(), 0),
			Transaction::read(PIN::channel(), 1023),
			Transaction::read(PIN::channel(), 800),
			Transaction::read(PIN::channel(), 600),
			Transaction::read(PIN::channel(), 400),
		];

		let mut adc = Mock::new(&expected);
		let mut keypad = KC11B04::new(PIN, MAP_10BIT);

		assert_eq!(
			(
				keypad.key_state(&mut adc),
				keypad.key_state(&mut adc),
				keypad.key_state(&mut adc),
				keypad.key_state(&mut adc),
				keypad.key_state(&mut adc)
			),
			(
				Ok(None),
				Ok(Some(Key::K4)),
				Ok(Some(Key::K3)),
				Ok(Some(Key::K2)),
				Ok(Some(Key::K1))
			)
		);
	}
}


src/mapping.rs
// Copyright 2023-2023 Robin van Boven
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

//! This module maps keys to their expected ADC readings.
//!
//! A specific mapping is represented as a [`KeyMap`].
//! Their values depend on the resolution of your ADC (8bit, 10bit, 12bit, etc.).
//! Several predefined mappings are exported in the root of the crate, such as [`MAP_10BIT`][crate::MAP_10BIT].
//! See the crate's [Constants][crate#constants] for this.
//!
//! ## Custom maps
//!
//! You can define custom maps. However floating point arithmetic is slow and inconsistent between different targets.
//! It's also *totally unnecessary to do at runtime* because we'll only ever compare integers against integers.
//!
//! You can of course do your own math ahead of time and write down literal integers.
//!
//! ```rust
//! use kc11b04::KeyMap;
//!
//! // The 10-bit KeyMap from the manufacturer's example table.
//! let map = KeyMap {
//! 	k1: 404,
//! 	k2: 607,
//! 	k3: 812,
//! 	k4: 1023,
//! 	margin: 30,
//! };
//! ```
//!
//! Or you could perform all floating point math in a constant expression.
//! Letting the compiler do this math *at compile time*, keeping the binary small and
//! avoiding the performance hit on your embedded device.
//!
//! ```rust
//! use kc11b04::KeyMap;
//!
//! // Floating point math is supported in constant definitions.
//! // And should "compile away" into just the final integers.
//! const CUSTOM_MAP: KeyMap<u16> = {
//! 	use kc11b04::mapping::{K1_F, K2_F, K3_F};
//! 	let max = 1023;
//! 	let margin = 0.03;
//! 	KeyMap {
//! 		k1: (max as f32 * K1_F) as u16,
//! 		k2: (max as f32 * K2_F) as u16,
//! 		k3: (max as f32 * K3_F) as u16,
//! 		k4: max,
//! 		margin: (max as f32 * margin) as u16,
//! 	}
//! };
//! ```
//!
//! This is effectively what the [`map_from_max!`][crate::map_from_max] macro allows you to do as well.
//!
//! ```rust
//! use kc11b04::{KeyMap, map_from_max};
//!
//! /// 10bit map, but with 15% margin.
//! const CUSTOM_MAP: KeyMap<u16> = map_from_max!(u16, 1023, 0.15);
//! ```
//!
//! ## Schematic and factors
//!
//! While the manufacturer provides a table of voltages, it's actually off slightly.
//! For 5V on `VCC` they list:
//! - K4 = 5V (100%)
//! - K3 = 4V (80%)
//! - K2 = 3V (60%)
//! - K1 = 2V (40%)
//!
//! This *roughly* lines up but not exactly.
//! Both my ADC and volt meter find slightly different values.
//! Taking a closer look at the schematic for the module:
//!
//! ![KC11B04 schematic][kc11b04-schema]
//!
//! We can see a voltage divider on the left based on the key being pressed.
//! However, there is also a pull-down resistor and a capacitor on the right side.
//! The simple table does not account of the pull-down resistor affecting the voltage divider.
//!
//! Ignoring the capacitor (which doesn't affect the stable state much) but adding the pull-down resistor `R5`,
//! the *equivalent resistors* become:
//!
//! - K1 = 3000 ohm : ~1961 ohm (`1/((1/2000) + (1/100000))`)
//! - K2 = 2000 ohm : ~2913 ohm (`1/((1/3000) + (1/100000))`)
//! - K3 = 1000 ohm : ~3846 ohm (`1/((1/4000) + (1/100000))`)
//! - K4 = 0 ohm : 100K ohm
//!
//! The formula then becomes `r2 / (r1 + r2)`.
//! For instance for K1: `~1961 / (3000 + ~1961) = ~0.395`.
//!
//! These factors are stored in `f32` resolution as constants.
//! And the new table becomes:
//!
//! | Key  | Constant | Fixed percent | Old percent |
//! | ---- | -------- | ------------- | ----------- |
//! | None | -        | 0%            | 0%          |
//! | K1   | [`K1_F`] | ~39.5%        | 40%         |
//! | K2   | [`K2_F`] | ~59.3%        | 60%         |
//! | K3   | [`K3_F`] | ~79.4%        | 80%         |
//! | K4   | -        | 100%          | 100%        |
//!
#![cfg_attr(
	feature = "doc-images",
	doc = ::embed_doc_image::embed_image!("kc11b04-schema", "docs/KC11B04-schema.svg")
)]
#![cfg_attr(
	not(feature = "doc-images"),
	doc = "[kc11b04-schema]: docs/KC11B04-schema.svg"
)]

use core::ops::{Add, Sub};

use crate::Key;

/// Maps keys to their expected ADC readings.
///
/// This depends on the ADC resolution, so you may want to use a constant for this such as [`MAP_10BIT`][crate::MAP_10BIT].
///
/// For creating a custom map, see the module documentation [`kc11b04::mapping`][crate::mapping].
#[cfg_attr(feature = "defmt-0-3", derive(defmt::Format))]
#[cfg_attr(feature = "ufmt-0-2", derive(ufmt::derive::uDebug))]
#[derive(Debug)]
pub struct KeyMap<Word> {
	/// The expected ADC reading for K1, before margins.
	///
	/// For predefined maps it's [`K1_F`] times the max reading of the ADC.
	pub k1: Word,

	/// The expected ADC reading for K2, before margins.
	///
	/// For predefined maps it's [`K2_F`] times the max reading of the ADC.
	pub k2: Word,

	/// The expected ADC reading for K3, before margins.
	///
	/// For predefined maps it's [`K3_F`] times the max reading of the ADC.
	pub k3: Word,

	/// The expected ADC reading for K4, before margins.
	///
	/// For predefined maps it's equal to the max reading of the ADC.
	pub k4: Word,

	/// The absolute margin a reading may deviate from the above expected values.
	/// The default is `3%` of the max reading of the ADC.
	pub margin: Word,
}

impl<Word> KeyMap<Word>
where
	Word: Copy + Add<Output = Word> + Sub<Output = Word> + Ord,
{
	/// Takes an ADC reading and finds whether it's in the expected range of a key.
	///
	/// Will be [`None`] when no key is pressed, but also for some simultaneous key combinations.
	pub fn key_from_reading(&self, val: Word) -> Option<Key> {
		match val {
			v if v > self.k1 - self.margin && v < self.k1 + self.margin => Some(Key::K1),
			v if v > self.k2 - self.margin && v < self.k2 + self.margin => Some(Key::K2),
			v if v > self.k3 - self.margin && v < self.k3 + self.margin => Some(Key::K3),
			v if v > self.k4 - self.margin => Some(Key::K4),
			_ => None,
		}
	}
}

/// Pull-down resistor value 100K ohms.
const R_DOWN: f32 = 100_000.0;

macro_rules! make_factor {
	($r1:literal, $r_rest:literal, $r_down:ident) => {{
		let r2 = 1.0 / ((1.0 / $r_rest) + (1.0 / $r_down));
		r2 / ($r1 + r2)
	}};
}

/// Relative factor for K1 button, `~39.5%` of the ADC's max reading.
///
/// See the module documentation [`kc11b04::mapping`][crate::mapping] for details.
pub const K1_F: f32 = make_factor!(3000.0, 2000.0, R_DOWN);

/// Relative factor for K2 button, `~59.3%` of the ADC's max reading.
///
/// See the module documentation [`kc11b04::mapping`][crate::mapping] for details.
pub const K2_F: f32 = make_factor!(2000.0, 3000.0, R_DOWN);

/// Relative factor for K3 button, `~79.4%` of the ADC's max reading.
///
/// See the module documentation [`kc11b04::mapping`][crate::mapping] for details.
pub const K3_F: f32 = make_factor!(1000.0, 4000.0, R_DOWN);

/// Defines a [`KeyMap`] based on the max reading of the ADC and optional margin factor.
///
/// The margin defaults to `0.03` (3% of the max parameter) if omitted.
///
/// ## Integer type
///
/// The macro and [`KeyMap`] take a `Word` generic parameter to support a variety of ADCs.
///
/// Keep in mind though this macro always maps the positive range, even when signed integers are used.
/// As negative voltages are out-of-spec for the KC11B04 module.
///
/// ```rust
/// # use kc11b04::{KeyMap, map_from_max};
/// /// Signed 11bit map. Meaning 10 bits of resolution in the positive voltage range,
/// /// so the max `1023` is the same as the 10 bit map.
/// const SIGNED_11BIT: KeyMap<i16> = map_from_max!(i16, 1023);
///
/// /// 24bit map needs a u32 to fit.
/// const UNSIGNED_24BIT: KeyMap<u32> = map_from_max!(u32, 0xFFFFFF);
/// ```
///
/// ## Performance note
///
/// This macro will use floating point math. It's strongly recommended to
/// **declare the result as a constant**, to avoid this FP math at runtime.
///
/// ```rust
/// # use kc11b04::{KeyMap, map_from_max};
/// /// 10bit map, but with 15% margin.
/// const CUSTOM_MAP: KeyMap<u16> = map_from_max!(u16, 1023, 0.15);
/// ```
#[macro_export]
macro_rules! map_from_max {
	($word:ident, $max:literal) => {
		map_from_max!($word, $max, 0.03)
	};
	($word:ident, $max:literal, $margin:literal) => {
		KeyMap {
			k1: ($max as f32 * $crate::mapping::K1_F) as $word,
			k2: ($max as f32 * $crate::mapping::K2_F) as $word,
			k3: ($max as f32 * $crate::mapping::K3_F) as $word,
			k4: $max,
			margin: ($max as f32 * $margin) as $word,
		}
	};
}

#[test]
fn read_10bit_samples() {
	// Use `i16` as it's less commonly used in examples.
	let map = crate::map_from_max!(i16, 1023, 0.03);
	assert_eq!(
		(
			map.key_from_reading(0),
			map.key_from_reading(300),
			map.key_from_reading(400),
			map.key_from_reading(500),
			map.key_from_reading(600),
			map.key_from_reading(700),
			map.key_from_reading(800),
			map.key_from_reading(900),
			map.key_from_reading(1023)
		),
		(
			None,
			None,
			Some(Key::K1),
			None,
			Some(Key::K2),
			None,
			Some(Key::K3),
			None,
			Some(Key::K4)
		)
	);
}


src/lib.rs
// Copyright 2023-2023 Robin van Boven
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

#![deny(unsafe_code)]
#![deny(missing_docs)]
#![allow(clippy::tabs_in_doc_comments)]
#![no_std]

//! Driver in Rust for the KC11B04 4-button analog keypad.
//!
//! A 3-pin module using a voltage divider circuit and analog pin to sense a key press.
//!
//! ![KC11B04 image][kc11b04-image]
//!
//! # Wiring notes
//!
//! The voltage on the `VCC` pin should correspond with the reference voltage (maximum value) your ADC is configured to read.
//!
//! For example the default on an Arduino Uno is that it matches the operating voltage of 5V.
//! But the ADC can be configured to use an external reference with the `AREF` pin. In this case `VCC` should match `AREF`.
//!
//! # Example
//!
//! ```rust
//! # use embedded_hal_mock::adc::*;
//! use kc11b04::{Key, KC11B04, MAP_10BIT};
//!
//! let mut adc = /* Configure your ADC using its HAL */
//! # Mock::new(&[Transaction::read(0, 1023)]);
//! let mut ad_pin = /* Set the pin connected to AD as analog input */
//! # MockChan0;
//!
//! // Create the keypad driver. Taking ownership of the pin,
//! // providing a map that matches the resolution of your ADC.
//! let mut keypad = KC11B04::new(ad_pin, MAP_10BIT);
//!
//! // Somewhere within loop { }
//! // Read current key state.
//! match keypad
//! 	.key_state(&mut adc)
//! 	.expect("Problem reading ADC channel")
//! {
//! 	Some(Key::K4) => { /* K4 key being pressed */ }
//! 	Some(Key::K3) => { /* K3 key being pressed */ }
//! 	Some(Key::K2) => { /* K2 key being pressed */ }
//! 	Some(Key::K1) => { /* K1 key being pressed */ }
//! 	None => { /* Either nothing, or multiple keys pressed */ }
//! }
//! ```
//!
//! # MSRV & Stability
//!
//! This crate compiles on stable Rust `1.60` and up. It _might_ still work on older rust versions,
//! but this isn't ensured. Upgrading the MSRV is considered SemVer breaking.
//!
#![cfg_attr(
	feature = "doc-images",
	doc = ::embed_doc_image::embed_image!("kc11b04-image", "docs/KC11B04.webp")
)]
#![cfg_attr(
	not(feature = "doc-images"),
	doc = "[kc11b04-image]: docs/KC11B04.webp"
)]

mod driver;
pub mod mapping;

pub use driver::*;
pub use mapping::KeyMap;

/// A named key on the [KC11B04][crate] module.
#[cfg_attr(feature = "defmt-0-3", derive(defmt::Format))]
#[cfg_attr(feature = "ufmt-0-2", derive(ufmt::derive::uDebug))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Key {
	/// `K1` key on the [KC11B04][crate] module.
	K1 = 1,
	/// `K2` key on the [KC11B04][crate] module.
	K2,
	/// `K3` key on the [KC11B04][crate] module.
	K3,
	/// `K4` key on the [KC11B04][crate] module.
	K4,
}

/// [`KeyMap`] for 8bit ADCs with a maximum reading of `255`.
pub const MAP_8BIT: KeyMap<u8> = map_from_max!(u8, 255);

/// [`KeyMap`] for 10bit ADCs with a maximum reading of `1023`.
pub const MAP_10BIT: KeyMap<u16> = map_from_max!(u16, 1023);

/// [`KeyMap`] for 12bit ADCs with a maximum reading of `4095`.
pub const MAP_12BIT: KeyMap<u16> = map_from_max!(u16, 4095);

/// [`KeyMap`] for signed 12bit ADCs with a maximum *positive* reading of `2047`.
pub const MAP_12BIT_SIGNED: KeyMap<i16> = map_from_max!(i16, 2047);

/// [`KeyMap`] for 16bit ADCs with a maximum reading of `65535`.
pub const MAP_16BIT: KeyMap<u16> = map_from_max!(u16, 65535);

/// [`KeyMap`] for signed 16bit ADCs with a maximum *positive* reading of `32767`.
pub const MAP_16BIT_SIGNED: KeyMap<i16> = map_from_max!(i16, 32767);

/// [`KeyMap`] for 24bit ADCs with a maximum reading of `16777215`.
pub const MAP_24BIT: KeyMap<u32> = map_from_max!(u32, 0xFFFFFF);
